dfc/tests/integration_test.go

```diff
@@ -0,0 +1,301 @@
+			m.numGetErrsAfter, maxNumGetErrs)
+	} else {
+		tlogf("Found %d GET errors (should be no more than %d) after target's local bucket map was updated",
+			m.numGetErrsAfter, maxNumGetErrs)
+	}
+}
+
+// 1. Unregister a target
+// 2. Create a local bucket
+// 3. Crash the primary proxy and PUT in parallel
+// 4. Failback to the original primary proxy, re-register target, and GET in parallel
+func TestProxyFailbackAndGetAndReRegisterInParallel(t *testing.T) {
+	var (
+		err error
+		num = 20000
+		m   = metadata{
+			t:               t,
+			delay:           15 * time.Second,
+			num:             num,
+			numGetsEachFile: 5,
+			repFilenameCh:   make(chan repFile, num),
+			semaphore:       make(chan struct{}, 10), // 10 concurrent GET requests at a time
+			wg:              &sync.WaitGroup{},
+		}
+		// Currently, a small percentage of GET errors can be reasonably expected as a result of this test.
+		// With the current design of dfc, there is exists a brief period in which the cluster map is synced to
+		// all nodes in the cluster during re-registering. During this period, errors can occur.
+		maxNumGetErrs = uint64(m.num * m.numGetsEachFile / 10) // 10 % of GET requests
+		filenameCh    = make(chan string, m.num)
+		errch         = make(chan error, 100)
+		sgl           *dfc.SGLIO
+		filesize      = uint64(1024)
+		seed          = int64(111)
+	)
+
+	// Step 1
+	m.smap, err = client.GetClusterMap(proxyurl)
+	checkFatal(err, t)
+	l := len(m.smap.Tmap)
+	if l < 2 {
+		t.Fatalf("Must have 2 or more targets in the cluster, have only %d", l)
+	}
+	m.origNumTargets = l
+	for m.sid = range m.smap.Tmap {
+		break
+	}
+	m.targetDirectURL = m.smap.Tmap[m.sid].DirectURL
+
+	err = client.UnregisterTarget(proxyurl, m.sid)
+	checkFatal(err, t)
+	n := len(getClusterMap(httpclient, t).Tmap)
+	tlogf("Unregistered target %s: the cluster now has %d targets\n", m.sid, n)
+
+	// Step 2
+	m.bucket = TestLocalBucketName
+	err = client.CreateLocalBucket(proxyurl, m.bucket)
+	checkFatal(err, t)
+
+	defer func() {
+		err = client.DestroyLocalBucket(proxyurl, m.bucket)
+		checkFatal(err, t)
+	}()
+
+	if usingSG {
+		sgl = dfc.NewSGLIO(filesize)
+		defer sgl.Free()
+	}
+
+	// Step 3
+	m.wg.Add(1)
+	go func() {
+		primaryCrash(t)
+		m.wg.Done()
+	}()
+
+	// PUT phase is timed to ensure it doesn't finish before primaryCrash() begins
+	time.Sleep(5 * time.Second)
+	tlogf("Putting %d files into bucket %s...\n", num, m.bucket)
+	putRandomFiles(0, seed, filesize, num, m.bucket, t, nil, errch, filenameCh, SmokeDir, SmokeStr, "", true, sgl)
+	selectErr(errch, "put", t, false)
+	close(filenameCh)
+
+	for f := range filenameCh {
+		m.repFilenameCh <- repFile{repetitions: m.numGetsEachFile, filename: f}
+	}
+	m.wg.Wait()
+
+	// Step 4
+	m.wg.Add(m.num*m.numGetsEachFile + 2)
+	go func() {
+		// Failback is timed to ensure it doesn't finish before GETs begin
+		time.Sleep(17 * time.Second)
```

Comment:
Interesting number of seconds. Can we avoid sleeps with random or hand-picked on certain machine number of seconds? As an idea: what if we share a channel between this function and doRepeatedGets(maybe we can do something for the sleep above). This function tries to read from channel and hangs until doRepeatedGets closes the channel and then continues. doRepeated can do a few GETs, then close channel and continue doing GETs, so you get two "sub-tests" reading a little % of data from stable cluster and the other reads go to the cluster while it is changing its config